lib/puppet/util.rb

```diff
@@ -25,17 +25,47 @@ module Util
 
   # Run some code with a specific environment.  Resets the environment back to
   # what it was at the end of the code.
-  def self.withenv(hash)
-    saved = ENV.to_hash
-    hash.each do |name, val|
-      ENV[name.to_s] = val
-    end
+  # Windows can store unicode chars in the environment, so need to use API calls
+  #   to get and set the environment instead of ENV due to encoding problems.
+  def self.withenv(hash, mode=:posix)
 
-    yield
-  ensure
-    ENV.clear
-    saved.each do |name, val|
-      ENV[name] = val
+    if (mode == :windows)
+      begin
+        require 'puppet/util/windows/process'
+
+        saved = Puppet::Util::Windows::Process.get_environment_strings
+        hash.each do |name, val|
+          # TODO Is it safe to assume a name in this hash is not utf8? Need to use .to_s for symbols
```

Comment:
When passing a symbol, later on in the call stack (wide_string) it raises an error as the Symbol class doesn't have a '+' method.  I'll modify the wide_string function to convert the input as a string if it isn't one.